subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1251 @@
+		break;
+	case MSG_DATA:
+		r = received_data(dev_data, message->block_index, ept_addr);
+		break;
+	default:
+		/* Silently ignore other messages types. They can be used in future
+		 * protocol version.
+		 */
+		break;
+	}
+
+exit:
+	if (r < 0) {
+		LOG_ERR("Failed to receive, err %d", r);
+	}
+}
+
+/**
+ * Callback called when ICMsg is bound.
+ */
+static void bound(void *priv)
+{
+	const struct device *instance = priv;
+	struct backend_data *dev_data = instance->data;
+
+	/* Set flag that ICMsg is bounded and now, endpoint bounding may start. */
+	atomic_or(&dev_data->flags, FLAG_ICMSG_BOUNDED);
+	schedule_ept_bound_process(dev_data);
+}
+
+/**
+ * Open the backend instance callback.
+ */
+static int open(const struct device *instance)
+{
+	const struct icmsg_with_buf_config *conf = instance->config;
+	struct backend_data *dev_data = instance->data;
+
+	static const struct ipc_service_cb cb = {
+		.bound = bound,
+		.received = received,
+		.error = NULL,
+	};
+
+	LOG_DBG("Open instance 0x%08X, initiator=%d", (uint32_t)instance,
+		dev_data->is_initiator ? 1 : 0);
+	LOG_DBG("  ICMsg, TX %d at 0x%08X, RX %d at 0x%08X",
+		(uint32_t)conf->icmsg_config.tx_shm_size,
+		(uint32_t)conf->icmsg_config.tx_shm_addr,
+		(uint32_t)conf->icmsg_config.tx_shm_size,
+		(uint32_t)conf->icmsg_config.tx_shm_addr);
+	LOG_DBG("  TX %d blocks of %d bytes at 0x%08X, max allocable %d bytes",
+		(uint32_t)conf->tx.block_count,
+		(uint32_t)conf->tx.block_size,
+		(uint32_t)conf->tx.blocks_ptr,
+		(uint32_t)(conf->tx.block_size * conf->tx.block_count -
+			   BLOCK_HEADER_SIZE));
+	LOG_DBG("  RX %d blocks of %d bytes at 0x%08X, max allocable %d bytes",
+		(uint32_t)conf->rx.block_count,
+		(uint32_t)conf->rx.block_size,
+		(uint32_t)conf->rx.blocks_ptr,
+		(uint32_t)(conf->rx.block_size * conf->tx.block_count -
+			   BLOCK_HEADER_SIZE));
+
+	return icmsg_open(&conf->icmsg_config, &dev_data->icmsg_data, &cb,
+			  (void *)instance);
+}
+
+/**
+ * Endpoint send callback function (with copy).
+ */
+static int send(const struct device *instance, void *token, const void *msg, size_t len)
+{
+	struct backend_data *dev_data = instance->data;
+	struct ept_data *ept = token;
+	uint32_t alloc_size;
+	uint8_t *buffer;
+	int r;
+
+	/* Allocate the buffer. */
+	alloc_size = len;
+	r = alloc_tx_buffer(dev_data, &alloc_size, &buffer, K_FOREVER);
+	if (r < 0) {
+		return r;
+	}
+
+	/* Copy data to allocated buffer. */
+	memcpy(buffer, msg, len);
+
+	/* Send data message. */
+	return send_block(dev_data, MSG_DATA, ept->addr, r, len);
+}
+
+/**
+ * Backend endpoint registration callback.
+ */
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	struct backend_data *dev_data = instance->data;
+	struct ept_data *ept = NULL;
+	int ept_index;
+	int r = 0;
+
+	/* Reserve new endpoint index. */
+	ept_index = atomic_inc(&dev_data->flags) & FLAG_EPT_COUNT_MASK;
+	if (ept_index >= NUM_EPT) {
```

Comment:
1. I cannot move before atomic_inc, because it will no longer be "atomic".
2. IMO, assertion is the best solution here. The endpoint are not something dynamic. There is specific number of endpoints in the application, so if you reach the limit then it is not a runtime error. This is a wrong configuration and for this case assert is the best.